### PR TITLE
HDDS-9081. Handling Netty back pressure when streaming containers for replication.

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?
Replication sends messages faster than the network/receiver can consume. This results in many messages being queued by Netty and the amount allocated memory surges and results in OOME. 

The change in this PR eliminates messages backlog and memory surge for both pull and push, thus allowing multiple concurrent replications.

It can be optimized to fully utilize manual control to avoid unnecessary delay (or sleep), following these [examples](https://github.com/grpc/grpc-java/tree/v1.30.2/examples/src/main/java/io/grpc/examples/manualflowcontrol). I don't have data to say if it's worth it yet.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-9081

## How was this patch tested?

Tested locally and verified it eliminates direct memory surges during replication.

### CODE REVIEW

**Comment by kerneltime** [comment]
```
[INFO] Running org.apache.hadoop.ozone.container.replication.TestGrpcContainerUploader
Error:  Tests run: 3, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 1.942 s <<< FAILURE! - in org.apache.hadoop.ozone.container.replication.TestGrpcContainerUploader
Error:  org.apache.hadoop.ozone.container.replication.TestGrpcContainerUploader.successfulReplication  Time elapsed: 1.841 s  <<< ERROR!
java.lang.ClassCastException: org.apache.hadoop.ozone.container.replication.TestGrpcContainerUploader$1 cannot be cast to org.apache.ratis.thirdparty.io.grpc.stub.CallStreamObserver
```

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/GrpcContainerUploader.java [hunk 1]
```diff
@@ -60,8 +61,9 @@ public OutputStream startUpload(long containerId, DatanodeDetails target,
       StreamObserver<SendContainerRequest> requestStream = client.upload(
           new SendContainerResponseStreamObserver(containerId, target,
               callback));
-      return new SendContainerOutputStream(requestStream, containerId,
-          GrpcReplicationService.BUFFER_SIZE, compression) {
+      return new SendContainerOutputStream(
+          (CallStreamObserver<SendContainerRequest>) requestStream,
```
Comment (kerneltime) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/GrpcContainerUploader.java:65]: Why not change ```private static class SendContainerResponseStreamObserver
      implements StreamObserver<SendContainerResponse> {``` to be `CallStreamObserver` 
The casting seems odd.

Comment (duongkame) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/GrpcContainerUploader.java:65]: actually, we don't take the `SendContainerResponseStreamObserver` as the input for SendContainerOutputStream, but the StreamObserver given by gRPC. And `StreamObserver` is the type gRPC gives. Internally, it returns an instance that implements `CallStreamObserver`, the reduces the result type to `StreamObserver`

Comment (kerneltime) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/GrpcContainerUploader.java:65]: Please add a few lines of comments explaining the casts that are done.

Comment (duongkame) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/GrpcContainerUploader.java]: Done.

**Pull request merged**